[DOCS][101] Add BYO vectors ingestion tutorial

leemthompo opened the pull request:
# 👁️  [URL preview](https://elasticsearch_bk_115112.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/bring-your-own-vectors.html)


- Adds a new bite-sized tutorial to **Search your data > Semantic search**
- This is a toy example to learn syntax of ingesting a set of existing vectors. Tries to add enough links to relevant material for follow-up without too much cognitive overload.
- Don't want to overload with information about the knn search side of things, but still making sure users can get where they need to next if they wanna drill down.

leemthompo committed:
[DOCS] Add BYO vectors ingestion tutorial

leemthompo committed:
Del whitespace

leemthompo committed:
Comment out flakey test, update ids

kderusso commented on docs/reference/search/search-your-data/ingest-vectors.asciidoc:
```diff
@@ -0,0 +1,131 @@
+<titleabbrev>Bring your own vector embeddings</titleabbrev>
+++++
+
+This tutorial demonstrates how to index documents that already have dense vector embeddings into {es}.
```
Is it worth adding an example for `sparse_vector` embeddings here as well?

leemthompo replied:
I think it would be best to keep this tightly focused to dense vectors and investigate demand for sparse vector equivalent going forward.

kderusso commented on docs/reference/search/search-your-data/ingest-vectors.asciidoc:
```diff
@@ -0,0 +1,131 @@
+    "properties": {
+      "review_vector": {
+        "type": "dense_vector",
+        "dims": 8, <1>
```
We could technically omit some of these, as `dims` can be dynamically calculated.

leemthompo replied:
I'll mention in annotation

kderusso commented on docs/reference/search/search-your-data/ingest-vectors.asciidoc:
```diff
@@ -0,0 +1,131 @@
+PUT /amazon-reviews/_doc/1
+{
+  "review_text": "This product is lifechanging! I'm telling all my friends about it.",
+  "review_vector": [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8]
```
Maybe add a note here emphasizing that the size of the `review_vector` array is 8 matching the dims count?

leemthompo replied:
👍

kderusso commented on docs/reference/search/search-your-data/ingest-vectors.asciidoc:
```diff
@@ -0,0 +1,131 @@
+[[bring-your-own-vectors-search-documents]]
+=== Step 3: Search documents with embeddings
+
+Now you can query these document vectors using a <<knn-retriever,`knn` retriever>>.
```
Nice to see retriever examples! 🎉

kderusso commented on docs/reference/search/search-your-data/ingest-vectors.asciidoc:
```diff
@@ -0,0 +1,131 @@
+  }
+}
+----
+// TEST[skip:flakeyknnerror]
```
Why is this test flakey?

leemthompo replied:
I'll rerun and print the error message but the knn query failed for some reason

kderusso commented on docs/reference/search/search-your-data/ingest-vectors.asciidoc:
```diff
@@ -0,0 +1,131 @@
+}
+----
+// TEST[skip:flakeyknnerror]
+<1> In this toy example, we're sending a raw vector as the query text. In a real-world scenario, you'll need to generate vectors for queries using an embedding model.
```
```suggestion
<1> In this simple example, we're sending a raw vector as the query text. In a real-world scenario, you'll need to generate vectors for queries using an embedding model.
```

kderusso commented on docs/reference/search/search-your-data/ingest-vectors.asciidoc:
```diff
@@ -0,0 +1,131 @@
+
+This was a simple example to help you understand the syntax for indexing a set of existing embeddings into {es}.
+
+In this toy example, we're sending a raw vector for the query text.
```
```suggestion
In this simple example, we're sending a raw vector for the query text.
```

kderusso commented on docs/reference/search/search-your-data/ingest-vectors.asciidoc:
```diff
@@ -0,0 +1,131 @@
+In a real-world scenario you won't know the query text ahead of time.
+You'll need to generate vectors for queries, on the fly, using an embedding model.
+
+For this you'll need to deploy a text embedding model in {es} and use the <<knn-query-top-level-parameters,`query_vector_builder` parameter>>.
```
It's also legitimate to do this client side and just send the vectors in with the request.

leemthompo replied:
Absolutely 👍

kderusso reviewed (commented):
Nice work!

leemthompo committed:
Updates, address feedback suggestions

Added explanation for dims parameter
Separated single and bulk document indexing examples
Improved explanations and wording throughout
Added tip for beginners about semantic search
Mentioned client-side vector generation as an alternative

leemthompo committed:
Add semantic search SVG overview

jeffvestal commented on docs/reference/search/search-your-data/ingest-vectors.asciidoc:
```diff
@@ -0,0 +1,141 @@
+[TIP]
+====
+The `dense_vector` type supports quantization to reduce the memory footprint required when searching float vectors.
+Learn more about balancing performance and accuracy in <<dense-vector-quantization,Dense vector quantization>>.
```
Is it worth mentioning that we auto-quantize with `int8_hnsw` by default for `desnse_vector` field type?

leemthompo committed:
Mention default int8_hnsw quantization

elasticsearchmachine commented:
Pinging @elastic/es-docs (Team:Docs)

szabosteve reviewed (approved):
This is awesome, LGTM!

elasticsearchmachine commented:
## 💔 Backport failed
The backport operation could not be completed due to the following error:
```
An unexpected error occurred when attempting to backport this PR.
```

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 115112`

leemthompo commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/115576">](https://github.com/elastic/elasticsearch/pull/115576)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.3.0 -->
